File: paleo_workbench/interchange/path_safety.py
"""Path-safety primitives for package extraction and manifest resolution.

Everything that turns a *manifest string* or an *archive entry name* into a
filesystem path must go through this module. The policy is fail-closed:
suspicious input is rejected with :class:`UnsafePathError`, never sanitized
silently, so callers cannot accidentally "skip" a malicious entry and continue
with a half-open package.

Defended cases:

- ``../`` traversal and absolute-path injection (POSIX and Windows forms)
- symlinks (and other non-regular members) escaping the package root
- duplicate names after Unicode NFC normalization
- case-insensitive collisions (macOS/Windows filesystems)
- Windows reserved device names (CON, NUL, COM1, ...)
- overlong paths / path components
"""

from __future__ import annotations

import os
import re
import unicodedata
from pathlib import Path, PurePosixPath


class UnsafePathError(ValueError):
    """Raised when an entry name must be rejected (fail-closed)."""


MAX_COMPONENT_LEN = 200
MAX_RELATIVE_LEN = 900

_WINDOWS_RESERVED = {
    "CON", "PRN", "AUX", "NUL",
    *(f"COM{i}" for i in range(1, 10)),
    *(f"LPT{i}" for i in range(1, 10)),
}

# Control characters and Windows-independent troublemakers in component names.
_BAD_CHARS_RE = re.compile(r"[<>:\"|?*\x00-\x1f]")


def safe_relative_path(name: str, *, what: str = "entry") -> PurePosixPath:
    """Validate a package-relative POSIX path and return it.

    Raises :class:`UnsafePathError` for anything that is not a clean,
    traversal-free, collision-safe relative path.
    """
    if not name or not name.strip():
        raise UnsafePathError(f"{what}: empty path")
    if "\x00" in name:
        raise UnsafePathError(f"{what}: NUL byte in {name!r}")

    normalized = unicodedata.normalize("NFC", name)
    if normalized != name:
        # Reject rather than rewrite: two manifest entries must never differ
        # only by Unicode normalization form (silent duplicate names).
        raise UnsafePathError(f"{what}: non-NFC path {name!r}")

    pure = PurePosixPath(normalized.replace("\\", "/"))
    if pure.is_absolute() or normalized.startswith("/") or re.match(r"^[A-Za-z]:", normalized):
        raise UnsafePathError(f"{what}: absolute path not allowed: {name!r}")

    parts = pure.parts
    if not parts:
        raise UnsafePathError(f"{what}: no components in {name!r}")
    for part in parts:
        if part in (".", ".."):
            raise UnsafePathError(f"{what}: traversal component in {name!r}")
        if not part.strip():
            raise UnsafePathError(f"{what}: blank path component in {name!r}")
        if part != part.rstrip(" ."):
            # NTFS/FAT strip trailing dots/spaces: "file.txt." would silently
            # overwrite "file.txt" and "com1 " would target the COM1 device.
            raise UnsafePathError(f"{what}: trailing dot/space in {part!r}")
        if len(part) > MAX_COMPONENT_LEN:
            raise UnsafePathError(f"{what}: path component too long in {name!r}")
        stem = part.split(".")[0].upper()
        if stem in _WINDOWS_RESERVED:
            raise UnsafePathError(f"{what}: Windows reserved name {part!r}")
        if _BAD_CHARS_RE.search(part):
            raise UnsafePathError(f"{what}: reserved character in {part!r}")

    rel = "/".join(parts)
    if len(rel) > MAX_RELATIVE_LEN:
        raise UnsafePathError(f"{what}: path too long: {name!r}")
    return pure


def check_collision(name: PurePosixPath, seen_casefold: set[str], seen_nfc: set[str]) -> None:
    """Track a validated name; raise on NFC-duplicate or case-insensitive clash."""
    key = str(name)
    if key in seen_nfc:
        raise UnsafePathError(f"duplicate entry after normalization: {key!r}")
    folded = key.casefold()
    if folded in seen_casefold:
        raise UnsafePathError(f"case-insensitive collision: {key!r}")
    seen_nfc.add(key)
    seen_casefold.add(folded)


def ensure_within_root(root: Path, candidate: Path) -> Path:
    """Resolve *candidate* under *root* and fail-closed on escape or symlink.

    Returns the resolved path. Rejects when any resolved component stops
    being inside *root* (covers both ``..`` survivors and symlinked dirs).
    """
    root = root.resolve()
    candidate = Path(candidate)
    if candidate.is_symlink():
        raise UnsafePathError(f"symlink rejected: {candidate}")
    resolved = candidate.resolve()
    if resolved != root and root not in resolved.parents:
        raise UnsafePathError(f"path escapes package root: {candidate}")
    return resolved


def safe_members(archive, *, what: str = "archive") -> list[str]:
    """Return validated member names of a :class:`zipfile.ZipFile`.

    Every name goes through :func:`safe_relative_path` plus a casefold/NFC
    collision pass. Any failure raises :class:`UnsafePathError` — callers must
    abort the whole extraction, not skip the entry.
    """
    # Imported lazily so this module stays usable without zipfile consumers.
    from zipfile import ZipInfo

    seen_casefold: set[str] = set()
    seen_nfc: set[str] = set()
    names: list[str] = []
    for info in archive.infolist():
        if isinstance(info, ZipInfo) and (info.external_attr >> 16) & 0o170000 == 0o120000:
            raise UnsafePathError(f"{what}: symlink entry rejected: {info.filename!r}")
        pure = safe_relative_path(info.filename, what=what)
        check_collision(pure, seen_casefold, seen_nfc)
        names.append(str(pure))
    return names


def extract_archive(archive, dest_root: Path, *, what: str = "package") -> list[Path]:
    """Extract a validated zip archive into *dest_root*, fail-closed.

    Files only; directories are created implicitly. Any unsafe entry aborts
    the whole extraction *before* any file is written (validation first,
    writes second — a partially extracted malicious package is never left
    behind on failure).
    """
    import zipfile

    if not isinstance(archive, zipfile.ZipFile):
        raise TypeError("extract_archive expects an opened zipfile.ZipFile")
    names = safe_members(archive, what=what)
    dest_root = Path(dest_root)
    dest_root.mkdir(parents=True, exist_ok=True)
    written: list[Path] = []
    for info, name in zip(archive.infolist(), names):
        target = ensure_within_root(dest_root, dest_root / name)
        if info.is_dir():
            target.mkdir(parents=True, exist_ok=True)
            continue
        target.parent.mkdir(parents=True, exist_ok=True)
        with archive.open(info) as src, open(target, "wb") as dst:
            while True:
                chunk = src.read(1024 * 1024)
                if not chunk:
                    break
                dst.write(chunk)
        written.append(target)
    return written


def sanitize_filename(stem: str, fallback: str = "export") -> str:
    """Make an arbitrary string safe to use as a single file name."""
    cleaned = unicodedata.normalize("NFC", stem)
    cleaned = re.sub(r'[<>:"|?*\x00-\x1f/\\]', "_", cleaned).strip(" .")
    if not cleaned:
        cleaned = fallback
    if cleaned.split(".")[0].upper() in _WINDOWS_RESERVED:
        cleaned = f"_{cleaned}"
    return cleaned[:MAX_COMPONENT_LEN]


def is_reserved_or_unsafe(name: str) -> bool:
    """True when *name* would be rejected as a path component."""
    try:
        safe_relative_path(name)
        return False
    except UnsafePathError:
        return True


def os_replace_atomic(temp_path: Path, target_path: Path) -> None:
    """os.replace with a directory fsync so renames survive power loss."""
    os.replace(temp_path, target_path)
    fd = os.open(str(target_path.parent), os.O_RDONLY)
    try:
        os.fsync(fd)
    except OSError:
        pass
    finally:
        os.close(fd)

File: libs/interchange/src/path_safety.cpp
// Faithful port of paleo_workbench/interchange/path_safety.py (pure half).
// Check order inside safe_relative_path is load-bearing: it decides which
// Python message the caller sees, so do not reorder.

#include <pwb/interchange/path_safety.hpp>

#include <pwb/interchange/unicode.hpp>

#include <array>
#include <cstddef>
#include <stdexcept>
#include <vector>

namespace pwb::interchange {

namespace {

// Python str.strip()/isspace() whitespace (full Unicode set, incl. the
// C0 file/group/record/unit separators U+001C..001F).
bool is_python_whitespace(char32_t cp) {
    if (cp <= 0x20) {
        return cp == 0x09 || cp == 0x0A || cp == 0x0B || cp == 0x0C
            || cp == 0x0D || (cp >= 0x1C && cp <= 0x20);
    }
    return cp == 0x85 || cp == 0xA0 || cp == 0x1680
        || (cp >= 0x2000 && cp <= 0x200A) || cp == 0x2028 || cp == 0x2029
        || cp == 0x202F || cp == 0x205F || cp == 0x3000;
}

bool is_all_whitespace(std::string_view text) {
    for (const char32_t cp : utf8_to_codepoints(text)) {
        if (!is_python_whitespace(cp)) {
            return false;
        }
    }
    return true;
}

bool is_blank_component(std::string_view part) {
    return is_all_whitespace(part);
}

std::string_view rstrip_space_dot(std::string_view part) {
    while (!part.empty() && (part.back() == ' ' || part.back() == '.')) {
        part.remove_suffix(1);
    }
    return part;
}

bool is_windows_reserved_stem(std::string_view stem_upper) {
    static constexpr std::array<std::string_view, 22> kReserved = {
        "CON", "PRN", "AUX", "NUL", "COM1", "COM2", "COM3", "COM4", "COM5",
        "COM6", "COM7", "COM8", "COM9", "LPT1", "LPT2", "LPT3", "LPT4",
        "LPT5", "LPT6", "LPT7", "LPT8", "LPT9",
    };
    for (const auto& name : kReserved) {
        if (stem_upper == name) {
            return true;
        }
    }
    return false;
}

char ascii_upper(char ch) {
    return (ch >= 'a' && ch <= 'z') ? static_cast<char>(ch - 'a' + 'A') : ch;
}

bool has_reserved_character(std::string_view part) {
    for (const unsigned char ch : part) {
        if (ch < 0x20 || ch == '<' || ch == '>' || ch == ':' || ch == '"'
            || ch == '|' || ch == '?' || ch == '*') {
            return true;
        }
    }
    return false;
}

bool starts_with_drive(std::string_view normalized) {
    if (normalized.size() < 2) {
        return false;
    }
    const auto first = static_cast<unsigned char>(normalized[0]);
    const bool alpha = (first >= 'A' && first <= 'Z')
        || (first >= 'a' && first <= 'z');
    return alpha && normalized[1] == ':';
}

std::vector<std::string_view> split_path_parts(std::string_view posix) {
    // PurePosixPath semantics: empty and "." segments dropped, duplicates
    // collapse; ".." survives for the traversal check.
    std::vector<std::string_view> parts;
    std::size_t start = 0;
    while (start <= posix.size()) {
        const std::size_t slash = posix.find('/', start);
        const auto segment = posix.substr(
            start, slash == std::string_view::npos ? posix.size() - start
                                                   : slash - start);
        if (!segment.empty() && segment != ".") {
            parts.push_back(segment);
        }
        if (slash == std::string_view::npos) {
            break;
        }
        start = slash + 1;
    }
    return parts;
}

std::string_view stem_of(std::string_view part) {
    const std::size_t dot = part.find('.');
    return dot == std::string_view::npos ? part : part.substr(0, dot);
}

}  // namespace

std::string python_repr(std::string_view text) {
    // Python picks double quotes only when the text holds ' but not ".
    const bool single_quote = text.find('\'') != std::string_view::npos
        && text.find('"') == std::string_view::npos;
    const char quote = single_quote ? '"' : '\'';
    // unicodedata.isprintable() escapes Cc (handled byte-wise below), the
    // C1 range 0x80..0xA0, and Zs/Zl/Zp separators; Cf/Co/Cn pass through
    // raw (documented bound; oracle pool contains none).
    const auto escaped_separator = [](char32_t cp) {
        return cp == 0xA0 || cp == 0x1680 || (cp >= 0x2000 && cp <= 0x200A)
            || cp == 0x2028 || cp == 0x2029 || cp == 0x202F || cp == 0x205F
            || cp == 0x3000;
    };
    static constexpr char kHex[] = "0123456789abcdef";
    std::string out;
    out.push_back(quote);
    for (const char32_t cp : utf8_to_codepoints(text)) {
        switch (cp) {
            case U'\n': out += "\\n"; break;
            case U'\r': out += "\\r"; break;
            case U'\t': out += "\\t"; break;
            case U'\\': out += "\\\\"; break;
            default:
                if (cp == static_cast<char32_t>(quote)) {
                    out.push_back('\\');
                    out.push_back(quote);
                } else if (cp < 0x20 || (cp >= 0x7F && cp <= 0xA0)) {
                    // Python uses \xNN below 0x100 (controls, DEL, C1, \xa0).
                    out += "\\x";
                    out.push_back(kHex[(cp >> 4) & 0xF]);
                    out.push_back(kHex[cp & 0xF]);
                } else if (escaped_separator(cp)) {
                    if (cp <= 0xFFFF) {
                        out += "\\u";
                        for (int shift = 12; shift >= 0; shift -= 4) {
                            out.push_back(kHex[(cp >> shift) & 0xF]);
                        }
                    } else {
                        out += "\\U";
                        for (int shift = 28; shift >= 0; shift -= 4) {
                            out.push_back(kHex[(cp >> shift) & 0xF]);
                        }
                    }
                } else {
                    out.append(codepoints_to_utf8(std::u32string(1, cp)));
                }
        }
    }
    out.push_back(quote);
    return out;
}

std::string safe_relative_path(const std::string& name, const std::string& what) {
    if (name.empty() || is_all_whitespace(name)) {
        throw UnsafePathError(what + ": empty path");
    }
    if (name.find('\x00') != std::string::npos) {
        throw UnsafePathError(what + ": NUL byte in " + python_repr(name));
    }

    const std::string normalized = nfc_normalize_utf8(name);
    if (normalized != name) {
        // Reject rather than rewrite: two manifest entries must never differ
        // only by Unicode normalization form (silent duplicate names).
        throw UnsafePathError(what + ": non-NFC path " + python_repr(name));
    }

    const std::string posix = [&] {
        std::string replaced = normalized;
        for (char& ch : replaced) {
            if (ch == '\\') {
                ch = '/';
            }
        }
        return replaced;
    }();
    if ((!posix.empty() && posix.front() == '/') || starts_with_drive(normalized)) {
        throw UnsafePathError(what + ": absolute path not allowed: "
                              + python_repr(name));
    }

    const std::vector<std::string_view> parts = split_path_parts(posix);
    if (parts.empty()) {
        throw UnsafePathError(what + ": no components in " + python_repr(name));
    }
    for (const std::string_view part : parts) {
        if (part == "." || part == "..") {
            throw UnsafePathError(what + ": traversal component in "
                                  + python_repr(name));
        }
        if (is_blank_component(part)) {
            throw UnsafePathError(what + ": blank path component in "
                                  + python_repr(name));
        }
        if (part != rstrip_space_dot(part)) {
            // NTFS/FAT strip trailing dots/spaces: "file.txt." would silently
            // overwrite "file.txt" and "com1 " would target the COM1 device.
            throw UnsafePathError(what + ": trailing dot/space in "
                                  + python_repr(part));
        }
        if (codepoint_length(part) > kMaxComponentLen) {
            throw UnsafePathError(what + ": path component too long in "
                                  + python_repr(name));
        }
        std::string stem_upper;
        for (const char ch : stem_of(part)) {
            stem_upper.push_back(ascii_upper(ch));
        }
        if (is_windows_reserved_stem(stem_upper)) {
            throw UnsafePathError(what + ": Windows reserved name "
                                  + python_repr(part));
        }
        if (has_reserved_character(part)) {
            throw UnsafePathError(what + ": reserved character in "
                                  + python_repr(part));
        }
    }

    std::string rel;
    for (const std::string_view part : parts) {
        if (!rel.empty()) {
            rel.push_back('/');
        }
        rel.append(part);
    }
    if (codepoint_length(rel) > kMaxRelativeLen) {
        throw UnsafePathError(what + ": path too long: " + python_repr(name));
    }
    return rel;
}

void check_collision(const std::string& name,
                     std::set<std::string>& seen_casefold,
                     std::set<std::string>& seen_nfc) {
    if (seen_nfc.count(name) != 0) {
        throw UnsafePathError("duplicate entry after normalization: "
                              + python_repr(name));
    }
    const std::string folded = casefold_utf8(name);
    if (seen_casefold.count(folded) != 0) {
        throw UnsafePathError("case-insensitive collision: " + python_repr(name));
    }
    seen_nfc.insert(name);
    seen_casefold.insert(folded);
}

std::filesystem::path ensure_within_root(const std::filesystem::path& root,
                                         const std::filesystem::path& candidate) {
    std::error_code ec;
    const std::filesystem::path resolved_root =
        std::filesystem::weakly_canonical(root, ec);
    if (std::filesystem::is_symlink(candidate)) {
        throw UnsafePathError("symlink rejected: " + candidate.string());
    }
    const std::filesystem::path resolved =
        std::filesystem::weakly_canonical(candidate, ec);
    if (resolved != resolved_root) {
        // "root in resolved.parents": the resolved path must be strictly
        // inside the root (covers ".." survivors and symlinked dirs alike).
        // lexically_relative is component-based, so it is separator- and
        // platform-correct (the previous text-prefix check only recognized
        // '/' and rejected every nested path on Windows).
        const std::filesystem::path relative =
            resolved.lexically_relative(resolved_root);
        const std::string relative_text = relative.generic_string();
        const bool inside = !relative_text.empty()
            && relative_text != "."
            && relative_text.rfind("../", 0) != 0
            && relative_text != "..";
        if (!inside) {
            throw UnsafePathError("path escapes package root: "
                                  + candidate.string());
        }
    }
    return resolved;
}

std::string sanitize_filename(const std::string& stem, const std::string& fallback) {
    std::string cleaned = nfc_normalize_utf8(stem);
    std::string replaced;
    replaced.reserve(cleaned.size());
    for (const unsigned char ch : cleaned) {
        if (ch < 0x20 || ch == '<' || ch == '>' || ch == ':' || ch == '"'
            || ch == '|' || ch == '?' || ch == '*' || ch == '/' || ch == '\\') {
            replaced.push_back('_');
        } else {
            replaced.push_back(static_cast<char>(ch));
        }
    }
    cleaned = replaced;
    // strip(" .") — code-point level, but ' ' and '.' are single bytes.
    std::size_t begin = 0;
    std::size_t end = cleaned.size();
    while (begin < end && (cleaned[begin] == ' ' || cleaned[begin] == '.')) {
        ++begin;
    }
    while (end > begin && (cleaned[end - 1] == ' ' || cleaned[end - 1] == '.')) {
        --end;
    }
    cleaned = cleaned.substr(begin, end - begin);
    if (cleaned.empty()) {
        cleaned = fallback;
    }
    std::string stem_upper;
    for (const char ch : stem_of(cleaned)) {
        stem_upper.push_back(ascii_upper(ch));
    }
    if (is_windows_reserved_stem(stem_upper)) {
        cleaned = "_" + cleaned;
    }
    const std::u32string cps = utf8_to_codepoints(cleaned);
    if (cps.size() > kMaxComponentLen) {
        return codepoints_to_utf8(cps.substr(0, kMaxComponentLen));
    }
    return cleaned;
}

bool is_reserved_or_unsafe(const std::string& name) {
    try {
        safe_relative_path(name);
        return false;
    } catch (const UnsafePathError&) {
        return true;
    }
}

}  // namespace pwb::interchange
